### PR TITLE
Vector mode memory

### PR DESCRIPTION
**Pull request by tgymnich**

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -2589,27 +2615,22 @@ class AdjointGenerator
...
+            Builder2.CreateMemCpy(ddst, dstAlign, dsrc, srcAlign, new_size);
+        call->setAttributes(MTI.getAttributes());
+        call->setTailCallKind(MTI.getTailCallKind());
+      };
 
+      applyChainRule(Builder2, rule, ddst, dsrc);
       return;
     }
 
-    size_t size = 1;
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:2600]: Why was this moved?

File: enzyme/Enzyme/GradientUtils.cpp [hunk 2]
```diff
@@ -2878,29 +2887,48 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
...
+    auto rule = [&oval]() { return Constant::getNullValue(oval->getType()); };
+
+    return applyChainRule(oval->getType(), BuilderM, rule);
   }
 
   if (isConstantValue(oval)) {
     // NOTE, this is legal and the correct resolution, however, our activity
     // analysis honeypot no longer exists
-    return getNewFromOriginal(oval);
+    auto rule = [&]() { return getNewFromOriginal(oval); };
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:2929]: Might as well do the newFromOriginal query once, outside the chain rule

File: enzyme/Enzyme/GradientUtils.cpp [hunk 3]
```diff
@@ -2985,58 +3013,72 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
...
-            AllocaInst *antialloca = bb.CreateAlloca(
-                arg->getValueType(), arg->getType()->getPointerAddressSpace(),
-                nullptr, arg->getName() + "'ipa");
-            invertedPointers.insert(std::make_pair(
-                (const Value *)oval, InvertedPointerVH(this, antialloca)));
+            Type *allocaTy = arg->getValueType();
 
-            if (arg->getAlignment()) {
+            auto rule1 = [&]() {
+              AllocaInst *antialloca = bb.CreateAlloca(
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:3019]: It might be wise here to instead create a larger alloca for the entire shadow, rather than each piece

File: enzyme/Enzyme/GradientUtils.cpp [hunk 4]
```diff
@@ -3079,25 +3121,34 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
...
-            arg->getType()->getAddressSpace(), arg->isExternallyInitialized());
-        arg->setMetadata("enzyme_shadow",
-                         MDTuple::get(shadow->getContext(),
-                                      {ConstantAsMetadata::get(shadow)}));
+        auto ip = arg->getInitializer() ? invertPointerM(arg->getInitializer(),
+                                                         B, /*nullShadow*/ true)
+                                        : Constant::getNullValue(type);
+
+        auto rule = [&](Value *ip) {
+          auto shadow = new GlobalVariable(
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:3132]: Same here with potentially creating a larger global, rather than several individual globals

File: enzyme/Enzyme/GradientUtils.cpp [hunk 5]
```diff
@@ -3301,88 +3438,114 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
...
     IRBuilder<> bb(getNewFromOriginal(inst));
     Value *asize = getNewFromOriginal(inst->getArraySize());
-    AllocaInst *antialloca = bb.CreateAlloca(
-        inst->getAllocatedType(), inst->getType()->getPointerAddressSpace(),
-        asize, inst->getName() + "'ipa");
-    invertedPointers.insert(std::make_pair(
-        (const Value *)oval, InvertedPointerVH(this, antialloca)));
-    if (inst->getAlignment()) {
+
+    auto rule1 = [&]() {
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:3467]: Similar allocation comment here.

File: enzyme/Enzyme/AdjointGenerator.h (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:2600]: this was removed by accident.

File: enzyme/Enzyme/GradientUtils.cpp (hunk 3 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:3019]: Since we cannot judge the vectorisation of coalesced allocas right now, i would go with the most canonical form and keep it this way.

File: enzyme/Enzyme/GradientUtils.cpp (hunk 5 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:3467]: Since we cannot judge the vectorisation of coalesced allocas right now, i would go with the most canonical form and keep it this way.

File: enzyme/Enzyme/GradientUtils.cpp [hunk 6]
```diff
@@ -2860,10 +2860,19 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
   }
 
   if (isa<ConstantPointerNull>(oval)) {
+    if (width > 1) {
```
Comment (wsmoses) [inline enzyme/Enzyme/GradientUtils.cpp:2863]: Shouldn't apply chain rule work on a vector width of 1 regardless? Why have the if here?

**Review by wsmoses — APPROVED** [review]
Address remaining comments otherwise lgtm

File: enzyme/Enzyme/GradientUtils.cpp (hunk 6 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:2863]: it should work without the if

File: enzyme/Enzyme/GradientUtils.cpp (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp]: fixed

File: enzyme/Enzyme/GradientUtils.cpp (hunk 4 above)
Comment (tgymnich) [reply enzyme/Enzyme/GradientUtils.cpp:3126]: done

File: enzyme/Enzyme/AdjointGenerator.h [hunk 7]
```diff
@@ -2589,11 +2610,21 @@ class AdjointGenerator
...
-      call->setTailCallKind(MTI.getTailCallKind());
+      Value *memFunc = MTI.getCalledOperand();
+
+      auto rule = [&](Value *ddst, Value *dsrc) {
+        CallInst *call = dyn_cast<CallInst>(
+            Builder2.CreateCall(MTI.getFunctionType(), memFunc,
+                                {ddst, dsrc, new_size, isVolatile}));
+        call->setAttributes(MTI.getAttributes());
+        call->setMetadata(LLVMContext::MD_tbaa,
+                          MTI.getMetadata(LLVMContext::MD_tbaa));
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:2621]: Also tbaa_struct (we use that as well, so this is needed if nesting AD).

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:2621]: done
